cep47-logic/src/lib.rs

```diff
@@ -15,6 +16,34 @@ pub mod tests;
 pub type TokenId = String;
 pub type Meta = BTreeMap<String, String>;
 
+pub trait PossiblyAddressable {
+    fn is_addressable(&self) -> bool;
+    fn matches_caller(&self, caller: &CallStackElement) -> bool;
+}
+
+impl PossiblyAddressable for casper_types::Key {
+    fn is_addressable(&self) -> bool {
+        match self {
+            Key::Account(_) => true,
+            Key::Hash(_) => true,
+            _ => false,
+        }
+    }
+
+    fn matches_caller(&self, caller: &CallStackElement) -> bool {
+        if !self.is_addressable() {
+            return false;
+        }
+        match (self, caller) {
+            (Key::Account(account_hash),  CallStackElement::Session { account_hash: account_hash_caller}) =>
+                account_hash == account_hash_caller,
+            (Key::Hash(hash_addr), CallStackElement::StoredContract { contract_package_hash: _, contract_hash: contract_hash_addr_caller}) =>
```

Comment:
ContractPackageHash should be extracted here, not ContractHash.

Reply:
I've also experimented with this feature and my version looks like that:
```
fn get_caller_id() -> Key {
    match runtime::get_call_stack().first().unwrap_or_revert() { // first or last?
        CallStackElement::Session { account_hash } => 
            account_hash_to_id(account_hash),
        CallStackElement::StoredSession { account_hash, contract_package_hash: _, contract_hash: _ } => // can stored contract call stored session? 
            account_hash_to_id(account_hash),
        CallStackElement::StoredContract { contract_package_hash, contract_hash: _ } =>
            contract_package_hash_to_id(contract_package_hash)
    }
}

fn account_hash_to_id(account_hash: &AccountHash) -> Key {
    (*account_hash).into()
}

fn contract_package_hash_to_id(hash: &ContractPackageHash) -> Key {
    (*hash).into()
}
```

Reply:
If tokens are to be controlled by a contract, it seems like it should be some definite piece of code, not just a pointer to a collection of contracts?

In your example, by the way, I believe the recovery of the calling entity is incorrect - as far as @goral09 and I could tell, it should be the next-to-last element in the stack, not first (or last - that would be the NFT contract itself, I think).